homeassistant/components/vacuum/demo.py

```diff
@@ -204,3 +212,111 @@ def send_command(self, command, params=None, **kwargs):
         self._status = 'Executing {}({})'.format(command, params)
         self._state = True
         self.schedule_update_ha_state()
+
+
+class StateDemoVacuum(StateVacuumDevice):
+    """Representation of a demo vacuum supporting states."""
+
+    def __init__(self, name, supported_features):
+        """Initialize the vacuum."""
+        self._name = name
+        self._supported_features = supported_features
+        self._state = STATE_DOCKED
+        self._fan_speed = FAN_SPEEDS[1]
+        self._cleaned_area = 0
+        self._battery_level = 100
+
+    @property
+    def name(self):
+        """Return the name of the vacuum."""
+        return self._name
+
+    @property
+    def should_poll(self):
+        """No polling needed for a demo vacuum."""
+        return False
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return self._supported_features
+
+    @property
+    def state(self):
+        """Return the current state of the vacuum."""
+        return self._state
+
+    @property
+    def battery_level(self):
+        """Return the current battery level of the vacuum."""
+        if self.supported_features & SUPPORT_BATTERY == 0:
+            return
+
+        return max(0, min(100, self._battery_level))
+
+    @property
+    def fan_speed(self):
+        """Return the current fan speed of the vacuum."""
+        if self.supported_features & SUPPORT_FAN_SPEED == 0:
+            return
+
+        return self._fan_speed
+
+    @property
+    def fan_speed_list(self):
+        """Return the list of supported fan speeds."""
+        assert self.supported_features & SUPPORT_FAN_SPEED != 0
+        return FAN_SPEEDS
+
+    @property
+    def device_state_attributes(self):
+        """Return device state attributes."""
+        return {ATTR_CLEANED_AREA: round(self._cleaned_area, 2)}
+
+    def start_pause(self, **kwargs):
+        """Start, pause or resume the cleaning task."""
+        if self.supported_features & SUPPORT_PAUSE == 0:
+            return
+
+        if self._state == STATE_CLEANING:
+            self._state = STATE_PAUSED
+        else:
+            self._state = STATE_CLEANING
+            self._cleaned_area += 1.32
+            self._battery_level -= 1
+        self.schedule_update_ha_state()
+
+    def stop(self, **kwargs):
+        """Stop the cleaning task, do not return to dock."""
+        if self.supported_features & SUPPORT_STOP == 0:
+            return
+
+        self._state = STATE_IDLE
+        self.schedule_update_ha_state()
+
+    def return_to_base(self, **kwargs):
+        """Return dock to charging base."""
+        if self.supported_features & SUPPORT_RETURN_HOME == 0:
+            return
+
+        self._state = STATE_RETURNING
```

Comment:
How this device can back to dock?

You can mock this behavior by using `hass.loop.call_later(30, set_state_to_dock)`